### PR TITLE
DrivAerML: polynomial LR decay (no cosine annealing)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Every DrivAerML experiment uses CosineAnnealingLR. The cosine schedule's periodic LR peaks are the root cause of most divergence failures — T_max deviations, multi-revisit, depth reduction, and symmetry augmentation all diverge at cosine LR peaks.

A monotonic LR decay (polynomial or linear) eliminates the periodic peaks entirely. The LR starts at 5e-4 and smoothly decays to near-zero, never revisiting high LR territory. This trades the basin-hopping benefits of cosine cycling for gradient stability.

If the champion's success at T_max=30 is mainly about spending most training time at LOW LR (the cosine trough), then a monotonic decay that keeps LR low for even longer should match or beat it.

## Instructions

Replace `CosineAnnealingLR` with `PolynomialLR` in train.py. Run two decay profiles:

**Run 1: Linear decay (power=1.0)**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --lr-schedule polynomial \
  --lr-poly-power 1.0 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group dm-polynomial-lr
```

**Run 2: Quadratic decay (power=2.0)**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --lr-schedule polynomial \
  --lr-poly-power 2.0 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group dm-polynomial-lr
```

Implementation: `torch.optim.lr_scheduler.PolynomialLR(optimizer, total_iters=999, power=X)`. Add `--lr-schedule` flag (cosine|polynomial) and `--lr-poly-power` flag. If --lr-schedule=cosine, use current CosineAnnealingLR. If polynomial, use PolynomialLR. No --grad-clip, no --weight-decay. Report best val_primary/surface_rel_l2_pct.

## Baseline

- **val_primary/surface_rel_l2_pct:** 3.997% at epoch 467
- **Config:** 4L/512d/8H, AdamW lr=5e-4, CosineAnnealingLR T_max=30, no-EMA, Fourier
- **W&B run:** bht6h42t
- **Reproduce:** `cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nobara:

## Results

### DrivAerML: Polynomial LR decay — NEGATIVE RESULT

| Run | Power | Best val `surface_rel_l2_pct` | Best epoch | Status | W&B |
|-----|-------|-------------------------------|------------|--------|-----|
| Linear | 1.0 | **17.88%** | ep64 | Catastrophic divergence ep69 (grad norm 13,967) | `i1qbv9t4` |
| Quadratic | 2.0 | **12.20%** | ep75 | Repeated instability, grad norms >100 at ep80/89/91 | `kkuery47` |
| **Baseline (cosine T_max=30)** | — | **3.997%** | ep467 | Stable | `bht6h42t` |

**Commands used:**
```bash
# Run 1: Linear (power=1.0) — GPU 0
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --lr-schedule polynomial --lr-poly-power 1.0 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --agent nobara --wandb-name "nobara/dm-poly-linear" --wandb-group dm-polynomial-lr

# Run 2: Quadratic (power=2.0) — GPU 1
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --lr-schedule polynomial --lr-poly-power 2.0 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --agent nobara --wandb-name "nobara/dm-poly-quadratic" --wandb-group dm-polynomial-lr
```

**Peak memory:** ~6 GB per run (GPU 0 and GPU 1)

### Implementation note

Added `--lr-schedule` (cosine|polynomial) and `--lr-poly-power` flags to train.py. PolynomialLR steps per-epoch (not per-batch like CosineAnnealingLR) to ensure smooth decay over the full training run. `total_iters=config.epochs`.

### What happened

**Run 1 (Linear, power=1.0):** The model showed repeated instability spikes from epoch 9 onwards (~every 10-15 epochs). At epoch 69, gradient norms exploded to 13,967, triggering catastrophic divergence. The model never recovered — grad norms remained in the 300-600 range with val oscillating 27-63% from ep70 onwards. Best result (17.88% at ep64) was achieved during a brief stable window between instability spikes.

**Run 2 (Quadratic, power=2.0):** More stable than linear but still showed periodic instability. Best result 12.20% at ep75. Major spikes at ep20 (33.6%), ep24 (54.0%), ep28 (65.9%), ep54 (30.4%), ep80 (31.5%), ep89 (36.1%), ep91 (70.8%). The quadratic curve keeps LR slightly lower at any given epoch, providing marginal stability improvement, but not enough.

**Key insight: The cosine schedule's periodic LOW-LR phases are load-bearing for stability, not just noise.** The hypothesis that cosine cycling only succeeds because it spends most time at low LR was **falsified** — polynomial decay keeps LR even higher for longer, and the model cannot handle it. Without grad-clip, LR=5e-4 is inherently unstable on DrivAerML beyond ~60-80 epochs. The cosine schedule with T_max=30 succeeds because:
1. The rapid cycling creates frequent LOW-LR recovery windows that heal gradient damage
2. The cycling acts as an implicit form of gradient stabilization
3. Basin-hopping from cosine peaks is actually beneficial (explores loss landscape)

Both polynomial runs are ~3-4x worse than baseline at their best, and both diverged before reaching the epochs where the baseline achieved its best (ep467).

### Suggested follow-ups
- **Polynomial + grad-clip=1.0** — the missing ingredient; grad-clip would prevent the catastrophic gradient explosions
- **Polynomial with much lower initial LR** (e.g. 1e-4 or 2e-4) — may be stable without grad-clip
- **Cosine with warm restarts (CosineAnnealingWarmRestarts)** — combines cosine cycling with longer decay periods

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3122 because ADVISOR: Both polynomial variants fail — linear: 17.88% then diverges ep69 (grad norm 13,967); quadratic: 12.20% then diverges ep91. Both 3-4x worse than the 3.833% baseline. Your analysis is the key takeaway: cosine's periodic low-LR troughs are LOAD-BEARING for stability — they act as implicit gradient-healing recovery windows. Polynomial decay keeps LR higher for longer and never provides these recovery troughs. Closing — polynomial LR is a dead end for DM..
